### PR TITLE
[libc++][memory] P2652R2: Disallow Specialization of `allocator_traits`

### LINKED ISSUE #105262: P2652R2: Disallow User Specialization of `allocator_traits`
**Link:** https://wg21.link/P2652R2

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements P2652R2 <https://wg21.link/P2652R2>:
- https://eel.is/c++draft/allocator.requirements.general
- https://eel.is/c++draft/memory.syn
- https://eel.is/c++draft/allocator.traits.general
- https://eel.is/c++draft/allocator.traits.members
- https://eel.is/c++draft/diff.cpp20.concepts
- https://eel.is/c++draft/diff.cpp20.utilities

### CODE REVIEW

File: libcxx/docs/ReleaseNotes/19.rst [hunk 1]
```diff
@@ -37,12 +37,16 @@ What's New in Libc++ 19.0.0?
...
+
 - P2637R3 - Member ``visit``
+- P2652R2 - Disallow User Specialization of ``allocator_traits``
 
 
 Improvements and New Features
 -----------------------------
-TODO
+
+- The ``_LIBCPP_ENABLE_CXX23_USER_SPECIALIZATION_OF_ALLOCATOR_TRAITS`` macro has been added to re-enable user specialization
```
Comment (philnik777) [inline libcxx/docs/ReleaseNotes/19.rst:48]: There is no need to add a flag for this. The feature has been added in C++23, and not many users will have started relying on it.

**Review by philnik777 — CHANGES_REQUESTED** [review]
Other than removing the flag this looks pretty good.

File: libcxx/docs/ReleaseNotes/19.rst (hunk 1 above)
Comment (H-G-Hristov) [reply libcxx/docs/ReleaseNotes/19.rst]: I thought so, removed it. Thank you for reviewing!

File: libcxx/utils/generate_feature_test_macro_components.py [hunk 2]
```diff
@@ -88,9 +88,9 @@ def add_version_header(tc):
         {
             "name": "__cpp_lib_allocate_at_least",
             "values": {
-                "c++23": 202106,
+                # "c++23": 202106,  # P0401R6 Providing size feedback in the Allocator interface
```
Comment (philnik777) [inline libcxx/utils/generate_feature_test_macro_components.py:91]: This line isn't needed anymore.

File: libcxx/include/__memory/allocation_result.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 23
+
+template <class _Pointer, class _SizeType = size_t>
```
[the remark this answers is not in the document]
Comment (Zingam) [reply libcxx/include/__memory/allocation_result.h:23]: To unbreak a circular dependancy.

File: libcxx/utils/generate_feature_test_macro_components.py (hunk 2 above)
Comment (Zingam) [inline libcxx/utils/generate_feature_test_macro_components.py:91]:
```suggestion
```

File: libcxx/include/__memory/allocation_result.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (H-G-Hristov) [reply libcxx/include/__memory/allocation_result.h]: I guess this can be folded into `allocator_traits.h`.

**Review by philnik777 — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**